src/lib/operators/join_index.cpp

```diff
@@ -201,40 +194,90 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
       }
     }
   } else {  // DATA JOIN since only inner joins are supported for a reference table on the index side
-    // Scan all chunks for index input
-    const auto chunk_count_index_input_table = _index_input_table->chunk_count();
-    for (ChunkID index_chunk_id{0}; index_chunk_id < chunk_count_index_input_table; ++index_chunk_id) {
-      const auto index_chunk = _index_input_table->get_chunk(index_chunk_id);
-      Assert(index_chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+    // Here we prefer to use table indexes if the join supports them. If no table index exists or other predicates than
+    // Equals or NotEquals are requested, chunk indexes are used. If no chunk index exists, NestedLoopJoin is used as a
+    // fallback solution.
+    const auto& table_indexes = _index_input_table->get_table_indexes(_adjusted_primary_predicate.column_ids.second);
```

Comment:
Copying the discussion from https://github.com/hyrise/hyrise/pull/2386#discussion_r697373795 to this PR:

>When columns are pruned during the query execution (by the `ColumnPruningRule`), a column id might not refer to the same column anymore. Imagine a table with the column ids {0, 1, 2, 3, 4, 5} and single-column indexes are created for the columns with the ids {2, 4}, resulting in two indexes. When the `ColumnPruningRule` prunes, for example, columns 0, 1, and 3, the remaining columns of the table cannot be identified by the initial column ids anymore.

>The following lines show the mapping from the initial column ids to the new ids, i.e., the resulting column ids after the column pruning:
0 -> (pruned)
1 -> (pruned)
2 -> 0
3 -> (pruned)
4 -> 1
5 -> 2

>The base table that is stored in the storage manager still contains all columns and `get_table_indexes` requires the column id to receive the single-column table indexes for that column. `_adjusted_primary_predicate.column_ids.second` in this line is a column id resulting after the column pruning. Thus, this code tries to retrieve table indexes with an updated column id (after the column pruning) using a function that expects the original column id (before the column pruning). Consequently, this can result in trying to retrieve table indexes from an incorrect column.

@tjjordan mentioned:
>To recreate this issue one need to run the TPCH benchmark with the --verify flag.
Failing benchmark querys: 7, 17, 19, 20, 21
I'm not sure if query 22 fails, the benchmark aborts just after.

Reply:
Regarding the column pruning issue:
As discussed in yesterday's Fortnightly, we want to add the pruned column ids or a column id mapping to the constructor of the index join. We can retrieve the pruned column ids from the corresponding StoredTableNode during LQP translation.